src/arm/radxa_rock_3c.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * Author: Nascs <nascs@radxa.com>
+ * Copyright (c) Radxa Limited.
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <mraa/common.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include "arm/radxa_rock_3c.h"
+#include "common.h"
+
+#define DT_BASE "/proc/device-tree"
+
+#define MAX_SIZE 64
+
+const char* radxa_rock_3c_serialdev[MRAA_RADXA_ROCK_3C_UART_COUNT] = { "/dev/ttyS2", "/dev/ttyS3", "/dev/ttyS4", "/dev/ttyS5", "/dev/ttyS9" };
+
+void
+mraa_radxa_rock_3c_pininfo(mraa_board_t* board, int index, int gpio_chip, int gpio_line, mraa_pincapabilities_t pincapabilities_t, char* pin_name)
+{
+
+    if (index > board->phy_pin_count)
+        return;
+
+    mraa_pininfo_t* pininfo = &board->pins[index];
+    strncpy(pininfo->name, pin_name, MRAA_PIN_NAME_SIZE);
+
+    if(pincapabilities_t.gpio == 1) {
+        pininfo->gpio.gpio_chip = gpio_chip;
+        pininfo->gpio.gpio_line = gpio_line;
+    }
+
+    pininfo->capabilities = pincapabilities_t;
+
+    pininfo->gpio.mux_total = 0;
+}
+
+mraa_board_t*
+mraa_radxa_rock_3c()
+{
+    mraa_board_t* b = (mraa_board_t*) calloc(1, sizeof(mraa_board_t));
+    if (b == NULL) {
+        return NULL;
+    }
+
+    b->adv_func = (mraa_adv_func_t*) calloc(1, sizeof(mraa_adv_func_t));
+    if (b->adv_func == NULL) {
+        free(b);
+        return NULL;
+    }
+
+    // pin mux for buses are setup by default by kernel so tell mraa to ignore them
+    b->no_bus_mux = 1;
+    b->phy_pin_count = MRAA_RADXA_ROCK_3C_PIN_COUNT + 1;
+
+    b->platform_name = PLATFORM_NAME_RADXA_ROCK_3C;
+    b->chardev_capable = 1;
+
+    // UART
+    b->uart_dev_count = MRAA_RADXA_ROCK_3C_UART_COUNT;
+    b->def_uart_dev = 0;
+    b->uart_dev[0].index = 2;
+    b->uart_dev[1].index = 3;
+    b->uart_dev[2].index = 4;
+    b->uart_dev[3].index = 5;
+    b->uart_dev[4].index = 9;
+
+    // I2C
+    b->i2c_bus_count = MRAA_RADXA_ROCK_3C_I2C_COUNT;
+    b->def_i2c_bus = 0;
+    b->i2c_bus[0].bus_id = 3;
+    b->i2c_bus[1].bus_id = 4;
+
+    // SPI
+    b->spi_bus_count = MRAA_RADXA_ROCK_3C_SPI_COUNT;
+    b->def_spi_bus = 0;
+    b->spi_bus[0].bus_id = 3;
+
+    // The PWM function cannot be used for the time being, because there is no way to use it one by one.
```

Comment:
ROCK 4怎么没有这个注释？

Reply:
// The PWM function cannot be used for the time being, because there is no way to use it one by one
这个忘记删了，这个 ken 之前是说 当时写这个 3c 的 mraa 的时候，pwm 无法使用来着

Reply:
如果现在3C的pwm可以正常工作的话就删掉